journal/src/main/java/io/zeebe/journal/file/record/JournalRecordSerializer.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.zeebe.journal.file.record;
+
+import org.agrona.DirectBuffer;
+import org.agrona.MutableDirectBuffer;
+
+public interface JournalRecordSerializer {
```

Comment:
I'm not sure about this interface:

1. Like I mentioned, I don't think having interfaces/implementation classes for the data to write is a good idea. It simplifies the method signature I guess, but then implementation also needs the impl classes? Why not have only impl classes? If it's only to simplify the signature, then for now I'd start by passing the fields directly in the various write methods.
2. We might want to consider later also passing the write buffer/read buffer + the offset/length to avoid allocating buffer views all the time. It's also a common pattern we use elsewhere and which the Agrona buffers support nicely.
3. Why not simply have `writeMetadata` and `writeData` as different methods? I'm not sure if the overload is useful here.
4. `getMetadataLength()` makes me realize we should think about what we want - is how the metadata/data laid out implementation details of the serializer, or is it the writer, and the serializer just serializes the messages to the given buffer? If it's implementation detail, then you just say `int length = serializer.writeData(buffer, index, asqn, data)` and it will already leave space for the metadata, and then `serializer.writeMetadata(buffer, checksum, length)` will write at the proper offset. Then the writer doesn't know if metadata is first, or after, etc, and we're coupling both more tightly I guess...idk, wdyt? But then you don't need `getMetadataLength()` for example. That also means with Kryo you wouldn't need to estimate size, I think?
5. I'm not sure if we need `hasMetadata` - maybe now, but later when we have the magic header, I guess we won't right?

Reply:
1. You are right. We don't need an interface and implementation classes for the data. Initially, we had different implementation for reading and writing. But that is not the case anymore. So it would make sense to only have the implementation classes.
2. Make sense. Do you want to address it in this PR?
3. Agree.
4. The writer and reader needs to know where the data is written so that it can calculate the checksum. If it is hidden in the serializer, we are back to the problem we discussed in the draft PR. 
5. When we have magic header we don't need `hasMetadata`.

Reply:
Re 4, you're right, I forgot - the reader needs to know where the offset of the data starts. I guess we need something like we have for the `LoggedEvent`, where we have `LoggedEvent#getValueBuffer()` and `LoggedEvent#getValueOffset()` (and same for the metadata). I guess this is fine for now, it still seems strange to me but maybe it's clearer what's the right API when we're mostly done.